reporting/src/main/java/org/wso2/testgrid/reporting/GraphDataProvider.java

```diff
@@ -0,0 +1,221 @@
+ *
+ * @since 1.0.0
+ */
+public class GraphDataProvider {
+    private TestPlanUOW testPlanUOW;
+    private static final int MAXIMUM_TIME_RANGE = 30;
+    private static final int TEST_EXECUTION_HISTORY_RANGE = 7;
+
+    public GraphDataProvider() {
+        this.testPlanUOW = new TestPlanUOW();
+    }
+
+    /**
+     * Provides the test failure summary for a given build job. Test Plan ids of the build job is retrieved by reading
+     * testgrid yaml files which contains in the current workspace.
+     *
+     * @param workspace   current workspace
+     * @throws ReportingException thrown when getting test plan ids by reading test plan yaml files
+     */
+    public List<BuildFailureSummary> getTestFailureSummary(String workspace) throws ReportingException {
+        List<TestCaseFailureResultDTO> testFailureSummary;
+        try {
+            testFailureSummary = testPlanUOW
+                    .getTestFailureSummary(FileUtil.getTestPlanIdByReadingTGYaml(workspace));
+            if (testFailureSummary.isEmpty()) {
+                return Collections.emptyList();
+            }
+            return processTestFailureSummary(testFailureSummary);
+        } catch (TestGridException e) {
+            throw new ReportingException("Error occurred while reading yaml files in the TG home", e);
+        } catch (TestGridDAOException e) {
+            throw new ReportingException("Error occurred while reading yaml files in the TG home", e);
+        }
+    }
+
+    /**
+     * Process the test failure summary which is retrieved by reading database and construct the list of
+     * {@link BuildFailureSummary} to draw test failure summary chart.
+     *
+     * @param testFailureSummary   list of {@link TestCaseFailureResultDTO} which are retrieved by quering the databse.
+     */
+    private List<BuildFailureSummary> processTestFailureSummary(List<TestCaseFailureResultDTO> testFailureSummary) {
+        TreeMap<String, BuildFailureSummary> testFailureSummaryMap = new TreeMap<>();
+        for (TestCaseFailureResultDTO testFailure : testFailureSummary) {
+            BuildFailureSummary buildFailureSummaryData = new BuildFailureSummary();
+            Gson gson = new GsonBuilder().create();
+            InfraCombination infraCombination = new InfraCombination();
+            String testName = testFailure.getName();
+            JsonElement jelem = gson.fromJson(testFailure.getInfraParameters(), JsonElement.class);
+            JsonObject jobj = jelem.getAsJsonObject();
+            infraCombination.setOs(StringUtil
+                    .concatStrings(jobj.get("OS").getAsString(), " - ", jobj.get("OSVersion").getAsString()));
+            infraCombination.setJdk(jobj.get("JDK").getAsString());
+            infraCombination.setDbEngine(StringUtil.concatStrings(jobj.get("DBEngine").getAsString(), " - ",
+                    jobj.get("DBEngineVersion").getAsString()));
```

Comment:
Shall we move strings like OS, JDK, DBEngine, and DBEngineVersion to constants? I believe these values will be used everywhere. Probably, we can do it later as an improvement.

Reply:
+1. We can keep these within reporting module. testgrid core is built in an abstract way. So, it does not know about these strings.